### PR TITLE
Add "Next Episode" button in the last 30 seconds of an episode

### PR DESCRIPTION
**Pull request by candry7731**
Adds "Next Episode" Button in the last 30 seconds of an episode. If use click the OK button on the remote the next episode will load.

**Features for next PR**
Add user custom setting to set time to show button
Add user custom setting to disable button
Add ability to prebuffer next episode to speed up load time

### CODE REVIEW

File: source/VideoPlayer.brs [hunk 1]
```diff
@@ -47,6 +47,8 @@ sub AddVideoContent(video, mediaSourceId, audio_stream_idx = 1, subtitle_idx = -
 
     if m.videotype = "Episode" or m.videotype = "Series"
         video.skipIntroParams = api_API().introskipper.get(video.id)
+        'print (meta.json.RunTimeTicks / 10000000) / 60
```
Comment (1hitsong) [inline source/VideoPlayer.brs]: Debug code

File: components/JFVideo.brs [hunk 2]
```diff
@@ -23,12 +23,25 @@ sub init()
...
+    m.moveDownnextEpisodeButtonAnimation = m.top.findNode("moveDownnextEpisodeButton")
 end sub
 
 '
 ' Checks if we have valid skip intro param data
 function haveSkipIntroParams() as boolean
-
+    'check current position
+    if int(m.top.position) >= (m.top.runTime - 30)
+        shownextEpisode()
```
Comment (1hitsong) [inline components/JFVideo.brs]: I'm not a fan of having nextEpisode code in the HaveSkipIntroParams(). Separation of concerns says SkipIntroParams has nothing to do with showing the Next Episode. I'd move it into its own thing.

**Review by 1hitsong — CHANGES_REQUESTED** [review]

File: components/JFVideo.brs (hunk 2 above)
Comment (candry7731) [reply components/JFVideo.brs]: I moved this out of the skip intro sub()

**Comment by 1hitsong** [comment]
@candry7731 If you can resolve the validate issue, I'll retest.

**Comment by candry7731** [comment]
Ill fix tonight

**Comment by 1hitsong** [comment]
I think the  `<JFButton id="nextEpisode"` component in the JFVideo.xml was accidentually removed. Clicking on any episode throws an error at `m.nextEpisodeButton = m.top.findNode("nextEpisode")` because the node isn't found.

**Comment by candry7731** [comment]
> I think the `<JFButton id="nextEpisode"` component in the JFVideo.xml was accidentually removed. Clicking on any episode throws an error at `m.nextEpisodeButton = m.top.findNode("nextEpisode")` because the node isn't found.

This is fixed.

**Comment by 1hitsong** [comment]
Button now appears!

At the end of an episode, once the button appeared, I went back on the trickplay bar to 2 minutes before the end and pressed OK expecting it to rewind, but instead it triggered the Next button.

**Comment by 1hitsong** [comment]
The last second of the episode before it changes shows the Next Episode button with a negative value.

image

**Comment by 1hitsong** [comment]
I rewound the video to the middle, but the Next Episode button stayed as it was at the last second. It should instead hide the button as it no longer applies this far back in the video.

image

**Comment by candry7731** [comment]
> The last second of the episode before it changes shows the Next Episode button with a negative value.
> 
> image

I could not reproduce this issue. all of my episodes stopped at 0

**Comment by candry7731** [comment]
> I rewound the video to the middle, but the Next Episode button stayed as it was at the last second. It should instead hide the button as it no longer applies this far back in the video.
> 
> image

This has been fixed. the button will now hide and restore focus to the video screen.

**Comment by 1hitsong** [comment]
Tried a different episode. It also has the -1 bug. I suppose add code so once it reaches 0 it doesn't countdown any further.

image

**Comment by 1hitsong** [comment]
This issue still exists: At the end of an episode, once the button appeared, I went back on the trickplay bar to 2 minutes before the end and pressed OK expecting it to rewind, but instead it triggered the Next button.

**Comment by candry7731** [comment]
what buttons did you use? the back arrow or the rewind button.

**Comment by 1hitsong** [comment]
1. Button shows at end of episode
2. Press LEFT on remote
3. Trickplay bar appears, continue pressing left until 2 minutes back
4. Press OK on remote

Expectation: Since I'm manipulating the trickplay bar, pressing OK would send the video back to that point.

Reality: Pressing OK triggers the Next Episode button

**Comment by candry7731** [comment]
Weird, I tested that exactly last night and had no issues with rewinding from the trickplay bar and pressing ok to restart the view..

I have the video set to focus if not within 30 seconds.... maybe need to add trickplay.focus to false also?
`sub handleNextEpisode()
    if int(m.top.position) >= (m.top.runTime - 30)
        shownextEpisode()
        updateCount()
    else
        m.nextEpisodeButton.visible = false
        m.nextEpisodeButton.setFocus(false)
        m.top.setFocus(true)
    end if`

**Comment by candry7731** [comment]
Added so more logic to the "ok" button press. Basically if the video is not playing the next up button is hidden and focus is returned to the video player

**Comment by alanazar** [comment]
Transcoding info (Up button) and subtitles (Down button) get stuck if you try pressing it during the 30sec countdown.
I even tested it at the start and it also gets stuck.
Only way to resolve it is to exit back to Roku then back in again
I checked the main app and it's working fine there so it's related to this PR

**Review by 1hitsong — APPROVED** [review]

**Comment by 1hitsong** [comment]
If there is no next episode (example: Last episode of last season) then don't show the button.

**Review by 1hitsong — CHANGES_REQUESTED** [review]
If there is no next episode (example: Last episode of last season) then don't show the button.

**Review by 1hitsong — APPROVED** [review]

**Comment by 1hitsong** [comment]
To be merged 11/22 unless an issue arises.

**Pull request merged**